### PR TITLE
New internal command Resolve-DbaComputerName that uses Resolve-DbaNetworkName to replace Resolve-NetBiosName

### LINKED ISSUE #5988: [Bug] Resolve-NetBiosName is unnecessary and breaks commands in complex envs
Due to it returning incompelte network names, whatever uses `Resolve-NetBiosName` should either trust the user and use `.ComputerName` or `Resolve-DbaNetworkName` should be used in its place.

````
C:\github\dbatools\functions\Copy-DbaBackupDevice.ps1:101:        $sourceNetBios = $Source.ComputerName
C:\github\dbatools\functions\Copy-DbaBackupDevice.ps1:175:                $sourcepath = Join-AdminUnc $sourceNetBios
$sourceServer.BackupDevices[$deviceName].PhysicalLocation
C:\github\dbatools\functions\Copy-DbaCredential.ps1:203:        $sourceNetBios = Resolve-NetBiosName $sourceServer
C:\github\dbatools\functions\Copy-DbaCredential.ps1:209:            Invoke-Command2 -ComputerName $sourceNetBios -Credential $credential -ScriptBlock { Get-ItemProperty
-Path "HKLM:\SOFTWARE\" }
C:\github\dbatools\functions\Copy-DbaCredential.ps1:211:            Stop-Function -Message "Can't connect to registry on $source" -Target $sourceNetBios -ErrorRecord $_
C:\github\dbatools\functions\Copy-DbaDatabase.ps1:385:                    $s.remotefilename = Join-AdminUNC $sourceNetBios $s.physical
C:\github\dbatools\functions\Copy-DbaDatabase.ps1:432:                        $s.remotefilename = Join-AdminUNC $sourceNetBios $s.physical
C:\github\dbatools\functions\Copy-DbaDatabase.ps1:469:                    $s.remotefilename = Join-AdminUNC $sourceNetBios $s.physical
C:\github\dbatools\functions\Copy-DbaDatabase.ps1:766:        $sourceNetBios = $sourceServer.ComputerName
C:\github\dbatools\functions\Copy-DbaDatabase.ps1:904:                $remoteSourcePath = Join-AdminUNC $sourceNetBios (Get-SqlDefaultPaths -SqlInstance $sourceServer
-filetype data)
C:\github\dbatools\functions\Copy-DbaLinkedServer.ps1:262:        $sourceNetBios = Resolve-NetBiosName $sourceserver
C:\github\dbatools\functions\Copy-DbaLinkedServer.ps1:266:            Invoke-Command2 -Raw -Credential $Credential -ComputerName $sourceNetBios -ScriptBlock {
Get-ItemProperty -Path "HKLM:\SOFTWARE\" } -ErrorAction Stop
C:\github\dbatools\functions\Copy-DbaLinkedServer.ps1:268:            Stop-Function -Message "Can't connect to registry on $source." -Target $sourceNetBios -ErrorRecord $_
C:\github\dbatools\functions\Expand-DbaDbLogFile.ps1:211:            $sourcenetbios = Resolve-NetBiosName $server
C:\github\dbatools\functions\Expand-DbaDbLogFile.ps1:264:                            [object]$AllDrivesFreeDiskSpace = Get-DbaDiskSpace -ComputerName $sourcenetbios |
Select-Object Name, SizeInKB
C:\github\dbatools\functions\Export-DbaCredential.ps1:134:                        $sourceNetBios = Resolve-NetBiosName $server
C:\github\dbatools\functions\Export-DbaCredential.ps1:138:                            Invoke-Command2 -Raw -Credential $Credential -ComputerName $sourceNetBios
-ScriptBlock { Get-ItemProperty -Path "HKLM:\SOFTWARE\" } -ErrorAction Stop
C:\github\dbatools\functions\Export-DbaCredential.ps1:140:                            Stop-Function -Message "Can't connect to registry on $instance." -Target
$sourceNetBios -ErrorRecord $_
C:\github\dbatools\functions\Export-DbaLinkedServer.ps1:108:            $sourceNetBios = Resolve-NetBiosName $server
C:\github\dbatools\functions\Export-DbaLinkedServer.ps1:112:                Invoke-Command2 -Raw -Credential $Credential -ComputerName $sourceNetBios -ScriptBlock {
Get-ItemProperty -Path "HKLM:\SOFTWARE\" } -ErrorAction Stop
C:\github\dbatools\functions\Export-DbaLinkedServer.ps1:114:                Stop-Function -Message "Can't connect to registry on $instance." -Target $sourceNetBios
-ErrorRecord $_
C:\github\dbatools\internal\functions\Get-DecryptedObject.ps1:33:    $sourceNetBios = Resolve-NetBiosName $server
C:\github\dbatools\internal\functions\Get-DecryptedObject.ps1:40:        [byte[]]$entropy = Invoke-Command2 -Raw -Credential $Credential -ComputerName $sourceNetBios
-argumentlist $serviceInstanceId {
C:\github\dbatools\internal\functions\Get-DecryptedObject.ps1:52:        $serviceKey = Invoke-Command2 -Raw -Credential $Credential -ComputerName $sourceNetBios
-ArgumentList $smkbytes, $Entropy {
C:\github\dbatools\internal\functions\Get-DecryptedObject.ps1:92:            $connString = "Server=ADMIN:$sourceNetBios\$instance;Trusted_Connection=True;Pooling=false"
C:\github\dbatools\internal\functions\Get-DecryptedObject.ps1:132:        $results = Invoke-Command2 -ErrorAction Stop -Raw -Credential $Credential -ComputerName
$sourceNetBios -ArgumentList $connString, $sql {
C:\github\dbatools\internal\functions\Get-SqlFileStructure.ps1:25:    $sourcenetbios = Resolve-NetBiosName $sourceserver
C:\github\dbatools\internal\functions\Get-SqlFileStructure.ps1:55:                $s.remotefilename = Join-AdminUnc $sourcenetbios $s.physical
C:\github\dbatools\internal\functions\Get-SqlFileStructure.ps1:90:                $s.remotefilename = Join-AdminUnc $sourcenetbios $s.physical
C:\github\dbatools\internal\functions\Get-SqlFileStructure.ps1:112:            $s.remotefilename = Join-AdminUnc $sourcenetbios $s.physical
````

### PR DESCRIPTION
**Pull request by andreasjordan**
## Type of Change

 - [x] Bug fix (non-breaking change, fixes #5988 )
 - [ ] New feature (non-breaking change, adds functionality, fixes # )
 - [ ] Breaking change (effects multiple commands or functionality, fixes # )
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1`)
 - [ ] Adding code coverage to existing functionality
 - [ ] Pester test is included
 - [ ] If new file reference added for test, has is been added to github.com/sqlcollaborative/appveyor-lab ?
 - [ ] Unit test is included
 - [ ] Documentation
 - [ ] Build system

### CODE REVIEW

**Comment by andreasjordan** [comment]
Although still in an early state, I would like to have general feedback here, @potatoqualitee / @wsmelton / @niphlod / @ClaudioESSilva

**Comment by potatoqualitee** [comment]
Seems reasonable to me!

**Comment by niphlod** [comment]
+1

**Comment by andreasjordan** [comment]
What I don't like is that it is now not just one line but a block:

```
try {
	$resolved = Resolve-DbaNetworkName -ComputerName $Source -Credential $Credential -EnableException
	$sourceFullComputerName = $resolved.FullComputerName
} catch {
	Stop-Function -Message "Error occurred while resolving $Source" -Category ConnectionError -ErrorRecord $_ -Target $Source
	return
}
```

The core part are two lines: Using `Resolve-DbaNetworkName` to get a full object and then just putting a specific property into the final variable.

What about a new parameter to `Resolve-DbaNetworkName` to only get the needed property?
Advantage: We can configure (with a config value?) inside of `Resolve-DbaNetworkName` what property is to be returned. So I can configure to use FQDN instead of FullComputerName without changing code.

And do we need `-EnableException` and `try {} catch {}` or can `Resolve-DbaNetworkName` just silently (only verbose output to analyse problems) return the input ComputerName on failure?

These both together would reduce it back to one line like it was with Resolve-NetBiosName.

I will think about this over the weekend, but if it's the total wrong direction, please stop me...

**Comment by niphlod** [comment]
that sounds like a job for an internal function: lots of cons (refactoring, mostly, to avoid duped code) for a smaaaall upside (If I got correctly you wanna end up with `$sourceFullComputerName= Resolve-DbaNetworkName -ComputerName $Source -Credential $Credential -EnableException -giveMejustTheFullName;` ). 
Adding a param to Resolve-Dba.. isn't going to cut it, it's a public function with a wealth of good infos.

**Comment by potatoqualitee** [comment]
What about modifying Resolve-NetBiosName and the code inside is the block you want to avoid typing every time? 

```
try {
	$resolved = Resolve-DbaNetworkName -ComputerName $Source -Credential $Credential -EnableException
	$sourceFullComputerName = $resolved.FullComputerName
} catch {
	Stop-Function -Message "Error occurred while resolving $Source" -Category ConnectionError -ErrorRecord $_ -Target $Source
	return
}
```

**Comment by andreasjordan** [comment]
I tried to understand the gererall concept and therefor had a look at the other commands that use `Resolve-DbaNetworkName` and `Invoke-Command2`.

These external commands use `Resolve-DbaNetworkName`:
Copy-DbaCredential
Copy-DbaInstanceAudit 
Copy-DbaLinkedServer
Disable-DbaForceNetworkEncryption
Disable-DbaHideInstance
Enable-DbaForceNetworkEncryption
Enable-DbaHideInstance 
Expand-DbaDbLogFile
Export-DbaCredential 
Export-DbaLinkedServer
Get-DbaClientProtocol
Get-DbaComputerSystem 
Get-DbaExtendedProtection
Get-DbaFilestream
Get-DbaForceNetworkEncryption 
Get-DbaHideInstance
Get-DbaInstanceProtocol 
Get-DbaLocaleSetting
Get-DbaMemoryUsage
Get-DbaNetworkActivity
Get-DbaOperatingSystem
Get-DbaPowerPlan
Get-DbaRegServer
Get-DbaService
Get-DbaStartupParameter 
Get-DbaUptime
Install-DbaInstance
Move-DbaDbFile
New-DbaComputerCertificate
New-DbaComputerCertificateSigningRequest
Remove-DbaNetworkCertificate
Rename-DbaDatabase 
Set-DbaExtendedProtection
Set-DbaNetworkCertificate
Set-DbaPowerPlan
Test-DbaConnection
Test-DbaDiskAlignment
Test-DbaDiskAllocation
Test-DbaPowerPlan
Test-DbaSpn
Update-DbaInstance
Update-DbaServiceAccount

These internal commands use `Resolve-DbaNetworkName`:
Get-DecryptedObject
Get-SqlFileStructure
Invoke-ManagedComputerCommand 
Resolve-SqlIpAddress

These external commands use `Invoke-Comnmand2` but not `Resolve-DbaNetworkName` (but they should, don't they?):
Add-DbaComputerCertificate
Add-DbaPfDataCollectorCounter
Clear-DbaConnectionPool
Get-DbaClientAlias
Get-DbaComputerCertificate
Get-DbaFeature
Get-DbaManagementObject
Get-DbaMsdtc 
Get-DbaNetworkCertificate
Get-DbaPfAvailableCounter
Get-DbaPfDataCollectorSet
Get-DbaPrivilege
Get-DbaProductKey
Import-DbaPfDataCollectorSetTemplate
Invoke-DbaDbLogShipping
New-DbaClientAlias
Remove-DbaClientAlias
Remove-DbaComputerCertificate
Remove-DbaPfDataCollectorCounter
Remove-DbaPfDataCollectorSet 
Set-DbaPrivilege
Start-DbaPfDataCollectorSet
Stop-DbaPfDataCollectorSet
Test-DbaConnection
Test-DbaManagementObject 

These internal commands use `Invoke-Command2` but not `Resolve-DbaNetworkName` (which is not a problem, because they should already get a full computer name):
Get-SQLInstanceComponent
Initialize-CredSSP
Invoke-Program
Register-RemoteSessionConfiguration
Restart-WinRMService
Test-PendingReboot
Unregister-RemoteSessionConfiguration


I thought about moving the resolve into Invoke-Command2, but there are some commands that issue more than one Invoke-Command2 to the same computer - that would result in multiple resolves. So I don't want to go that way.

But as the different commands don't follow one pattern in using `Resolve-DbaNetworkName` we should maybe introduce a replacement for `Resolve-NetBiosName` (as Chrissy suggested) and change all these commands to use the replacement.

So I would like to talk with you about what the new internal commands has to do and what it should be called.

To be most easy to use, it should take a [DbaInstanceParameter] as ComputerName and optionally a [PSCredential] as Credential, and return something that can be used as a ComputerName in Invoke-Command2. As we only need the plain ComputerName, the command should return just a string that we can use in a verbose message and pass on to Invoke-Command2.
The command should then try to run `Resolve-DbaNetworkName`, in the event of a failure `Resolve-DbaNetworkName` with `-Turbo` and in the event of a failure just return `.ComputerName` from the input. The command should only output verbose messages about the failures to not bother the user. Or should we output warnings because if the resolving fails, the following `Invoke-Command2` would maybe fail also?
The default would be to return the resolved `.FullComputerName` but I think a config parameter would be nice to change that to something else like `.FQDN`. We already have the config parameter "commands.resolve-dbanetworkname.bypass" to disable the resolving completely.

So we need an name for the command and a name for the returned string in the commands that sue the command.
How about `Resolve-DbaComputerName` and `$resolvedComputerName`?

**Comment by andreasjordan** [comment]
Changed my mind about the config and added a parameter to return other properties than FullComputerName. 
Used the new command in some of the commands, so that you can have a look...

**Comment by andreasjordan** [comment]
I used the input $SqlInstance when possible and not the $server, as the $server.ComputerName is only .NetName of the SMO.

**Comment by potatoqualitee** [comment]
@niphlod and @wsmelton can you give updated feedback based on the changed requirements? Merci 🙇🏼

**Comment by andreasjordan** [comment]
For information: New internal command is Resolve-DbaComputerName, I used it in Copy-DbaCredential, Copy-DbaInstanceAudit and Copy-DbaLinkedServer - so you can have a look how I think it should be used.

**Comment by andreasjordan** [comment]
We should include the rule "do not add domain part for a local host" that helped with Install-DbaInstance (https://github.com/sqlcollaborative/dbatools/issues/6976#issuecomment-880562827 / #7564).

**Comment by andreasjordan** [comment]
Is this the right way? Then I would change the other commands touched here and open the pull request for review.

**Comment by potatoqualitee** [comment]
i like it. should we keep it until after 1.1.0 or include it?

**Comment by andreasjordan** [comment]
Does not have to be in 1.1.0 - just want to get this finished in the next days. Will now continue to work on this.

**Comment by potatoqualitee** [comment]
@niphlod  or @wsmelton can we get your detail oriented eyes to approve this PR?

File: functions/Copy-DbaLogin.ps1 [hunk 1]
```diff
@@ -222,7 +222,8 @@ function Copy-DbaLogin {
                 continue
             }
 
-            $serverName = Resolve-NetBiosName $sourceServer
+            # Here we don't need the FullComputerName, but only the maschine name to compare to the host part of the login name. So ComputerName should be fine.
```
Comment (niphlod) [inline functions/Copy-DbaLogin.ps1:225]: typo, machine instead of maschine

File: functions/Sync-DbaLoginPermission.ps1 [hunk 2]
```diff
@@ -106,7 +106,8 @@ function Sync-DbaLoginPermission {
                     continue
                 }
 
-                $serverName = Resolve-NetBiosName $sourceServer
+                # Here we don't need the FullComputerName, but only the maschine name to compare to the host part of the login name. So ComputerName should be fine.
```
Comment (niphlod) [inline functions/Sync-DbaLoginPermission.ps1:109]: maschine --> machine

**Review by niphlod — CHANGES_REQUESTED** [review]
apart from typos, seems fine.
I'm still not a fan of to Resolve-DbaComputerName and the -Property param but till it's internal there's no reason to be nitpicky.

File: functions/Copy-DbaLogin.ps1 (hunk 1 above)
Comment (andreasjordan) [reply functions/Copy-DbaLogin.ps1]: Thanks. Fixed.

File: functions/Sync-DbaLoginPermission.ps1 (hunk 2 above)
Comment (andreasjordan) [reply functions/Sync-DbaLoginPermission.ps1]: Thanks. Fixed.

**Comment by potatoqualitee** [comment]
Thank you 🙇🏼

**Review by potatoqualitee — APPROVED** [review]
approved per niph

**Comment by potatoqualitee** [comment]
Here we goooo!

**Pull request merged**